kong/resty/dns_client/utils.lua

```diff
@@ -0,0 +1,273 @@
+  resolv.search = resolv.search or (resolv.domain and { resolv.domain })
+
+  -- check if timeout is 0s
+  if resolv.options.timeout and resolv.options.timeout <= 0 then
+    log(NOTICE, "A non-positive timeout of ", resolv.options.timeout,
+                "s is configured in resolv.conf. Setting it to 2000ms.")
+    resolv.options.timeout = 2000 -- 2000ms is lua-resty-dns default
+  end
+
+  -- remove special domain like "."
+  if resolv.search then
+    for i = #resolv.search, 1, -1 do
+      if resolv.search[i] == "." then
+        table_remove(resolv.search, i)
+      end
+    end
+  end
+
+  -- nameservers
+  if resolv.nameserver then
+    local nameservers = {}
+
+    for _, address in ipairs(resolv.nameserver) do
+      local ip, port, t = utils.parseHostname(address)
+      if t == "ipv4" or
+        (t == "ipv6" and not ip:find([[%]], nil, true) and enable_ipv6)
+      then
+        table_insert(nameservers, port and { ip, port } or ip)
+      end
+    end
+
+    resolv.nameservers = nameservers
+  end
+  return resolv
+end
+
+
+function _M.is_fqdn(name, ndots)
+  if name:sub(-1) == "." then
+    return true
+  end
+  local _, dot_count = name:gsub("%.", "")
+  return (dot_count >= ndots)
+end
+
+
+-- construct names from resolv options: search, ndots and domain
+function _M.search_names(name, resolv, hosts)
+  if not resolv.search or _M.is_fqdn(name, resolv.ndots) then
+    return { name }
+  end
+
+  local names = {}
+  for _, suffix in ipairs(resolv.search) do
+    table_insert(names, name .. "." .. suffix)
+  end
+
+  if hosts and hosts[name] then
+    table_insert(names, 1, name)
+  else
+    table_insert(names, name)
+  end
+
+  return names
+end
+
+
+function _M.ipv6_bracket(name)
+  if name:match("^[^[].*:") then  -- not rigorous, but sufficient
+    return "[" .. name .. "]"
+  end
+  return name
+end
+
+
+-- util APIs to balance @answers
+
+function _M.get_next_round_robin_answers(answers)
+  answers.last = (answers.last or 0) % #answers + 1
+  return answers[answers.last]
+end
+
+
+-- based on the Nginx's SWRR algorithm and lua-resty-balancer
+local function swrr_next(answers)
+  local total = 0
+  local best = nil    -- best answer in answers[]
+
+  for _, answer in ipairs(answers) do
+    -- 0.1 gives weight 0 record a minimal chance of being chosen (rfc 2782)
+    local w = (answer.weight == 0) and 0.1 or answer.weight
+    local cw = answer.cw + w
+    answer.cw = cw
+    if not best or cw > best.cw then
+      best = answer
+    end
+    total = total + w
+  end
+
+  best.cw = best.cw - total
+  return best
+end
+
+
+local function swrr_init(answers)
+  for _, answer in ipairs(answers) do
+    answer.cw = 0   -- current weight
+  end
+
+  -- random start
+  for _ = 1, math_random(#answers) do
+    swrr_next(answers)
+  end
+end
+
+
+-- gather records with the lowest priority in SRV record
+local function filter_lowest_priority_answers(answers)
+  local lowest_priority = answers[1].priority
```

Comment:
```suggestion
  local lowest_priority = assert(answers[1].priority)
```

Reply:
Is this safe? there should be a priority field. If there is none, there is an underlying issue somewhere. But from Kong perspective, should we maybe log an error, ignore the answer, and try to keep chugging along instead of failing hard?

Not saying we should, just wondering. Opinions welcome.

Reply:
I think we dont need to worry this, I have reviewed the code of `lua-resty-dns`, if the answers type is SRV, it MUST have prioirty field. 

And otherwise, there are so many fields needs us to check like `cname` `target` `priority` `weight` ...

Reply:
Could we add a comment to clarify this concern?

Reply:
comment added